django/thunderstore/social/providers.py

```diff
@@ -0,0 +1,112 @@
+from typing import Dict
+
+import requests
+from django.conf import settings
+
+
+class BaseOauthHelper:
+    token = None
+
+    def __init__(self, code: str, redirect_url: str) -> None:
+        self.code = code
+        self.redirect_url = redirect_url
+
+    def complete_login(self) -> None:
+        """
+        Fetch access token with the temporary code and configured secret.
+        """
+        data = {
+            "client_id": self.CLIENT_ID,
```

Comment:
here `self.CLIENT_ID` and `self.CLIENT_SECRET` aren't declared, which will lead to a type checking error (if we were running a type checker). You should be able to declare them as required properties somehow within the abstract base class without defining them there, but again I can't remember how that worked out in practice.

Reply:
Implemented by combining `@abstractmethod` and `@property`.